GVFS/GVFS.Common/NamedPipes/LockNamedPipeMessages.cs

```diff
@@ -317,23 +320,46 @@ internal static LockData FromBody(string body)
 
                     // ParsedCommandLength should be the length of the string at the end of the message
                     // Add the length of the previous parts, plus delimiters
-                    int startingSpot = dataParts[0].Length + dataParts[1].Length + dataParts[2].Length + dataParts[3].Length + 4;
-                    if ((startingSpot + parsedCommandLength) != body.Length)
+                    int commandStartingSpot = dataParts[0].Length + dataParts[1].Length + dataParts[2].Length + dataParts[3].Length + 4;
+                    if ((commandStartingSpot + parsedCommandLength) >= body.Length)
                     {
                         throw new InvalidOperationException(string.Format("Invalid lock message. The parsedCommand is an unexpected length, got: {0} from message: '{1}'", parsedCommandLength, body));
                     }
 
-                    parsedCommand = body.Substring(startingSpot, parsedCommandLength);
+                    parsedCommand = body.Substring(commandStartingSpot, parsedCommandLength);
+
+                    string sessionIdSubString = body.Substring(commandStartingSpot + parsedCommandLength + 1);
+                    string[] sessionIdParts = sessionIdSubString.Split(MessageSeparator);
+                    if (!int.TryParse(sessionIdParts[0], out int sessionIdLength))
+                    {
+                        throw new InvalidOperationException(string.Format("Invalid lock message. Expected session id length, got: {0} from message: '{1}'", sessionIdParts[0], body));
+                    }
+
+                    int sessionIdStartingSpot = commandStartingSpot + parsedCommandLength + sessionIdParts[0].Length + 2;
+                    if ((sessionIdStartingSpot + sessionIdLength) != body.Length)
+                    {
+                        throw new InvalidOperationException(string.Format("Invalid lock message. The sessionId is an unexpected length, got: {0} from message: '{1}'", sessionIdLength, body));
+                    }
+
+                    string sessionId = body.Substring(sessionIdStartingSpot, sessionIdLength);

```

Comment:
I'm having a little trouble following the above, but I think it is expecting a "%d %s" past the end of the previous
message for the SID-length and the SID-string itself.  I think this is fine as long as we treat the content of the
SID-string as an opaque string and not try to parse the content of it.

Reply:
Yes that is right and I don't like all the formatting and parsing that we are doing ourselves here but it is basically a `|` delimited string but there are strings that are length "prefixed" so that the string itself can have `|`s in it.  So it is something like:
`123|true|false|13|parsedCommand|9|sessionId`

I will add a comment on the format so that this is a little easier to understand.